### PR TITLE
Update transient.el.  Don't intern transient symbols

### PR DESCRIPTION
**Pull request by joaotavora**
See Emacs bug#68568

> Please use a dedicated feature branch for your feature request, instead of asking us to merge "your-fork/master" into the 
> "origin/master".  The use of dedicated branches makes it much more convenient to deal with pull-requests, especially when > using Magit to do so.

I don't use Magit.  This is a just a fix for a very annoying bug where completion gets completely broken just because I use a package that uses transient.

> If you were about to open a pull-request asking us to merge your version of "master", then see [1] for instructions on how to quickly fix that and some information on why we ask you to do so.

YOu can push to my master of magit any time.  The patch is trivial.

### CODE REVIEW

**Comment by psionic-k** [comment]
I played with a backport to 0.5.2 and on master.  I still see all of the commands in my completions.  If that's the motivation for change, what are we doing?

```elisp
(intern-soft "transient:tsc-cowsay:-b") ; transient:tsc-cowsay:-b
```
So the interning is happening anyway.

I'm not sure that an uninterned symbol can be used in a keymap, and if not, then the only implementations use private command names like these or re-implement everything on top repeated calls to `read-key`.

**Comment by joaotavora** [comment]
> I played with a backport to 0.5.2 and on master. I still see all of the commands in my completions. If

Likely the commands were already there from previous internations.  

Try the patch from scratch or use this to unintern the existing ones:

```
(mapatoms (lambda (s) (when (string-match "^transient--:" (symbol-name s)) (unintern s))))
```

**Comment by psionic-k** [comment]
I definitely did not just check for symbols on dirty Emacsen.  You want to look at your obarray after loading GPTel etc, which just defines the prefixes.

**Comment by joaotavora** [comment]
You're right, this is not enough for all symbols.  It is somehow enough for the really long ones

**Comment by joaotavora** [comment]
Definitely don't set any options in user Emacs.   I don't want to set that option: and I think it's perfectly legitimate to not want to and still be able to use completion elsewhere without monstrous useless transient commands following me around.

So could you explain why those symbols need to be interned in the obarray: (1) at compile-time (2) at load time (3) at run-time.

Also could you explain why some symbols need to have the actual description in their symbol names and why something like `gensym` couldn't be used instead?

**Comment by tarsius** [comment]
> Definitely don't set any options in user Emacs. I don't want to set that option:

I think you misunderstand what my predicate does. It excludes only those symbols that you don't want included. (Okay okay, it might also exclude symbols with non-nil `completion-predicate`, but that was a misguided last minute addition in my initial draft, just before posting it here, and I have already removed it again.)

Right now you only have a choice between setting that option to

```elisp
(defun transient-command-completion-non-suffix-only-p (symbol _buffer)
  "Say whether SYMBOL should be offered as a completion.
If SYMBOL is only intended to be called as suffix command from
a transient prefix commands, return nil, t otherwise.  Consider
using `command-completion-default-include-p' instead."
  (not (eq (car (command-modes symbol)) 'not-a-mode)))
```

, live with the noise, or stop using transient.

(Choosing the first option will achieve your goal *(and nothing else)*, so I hope you will reconsider.)

> So could you explain why those symbols need to be interned in the obarray

Yes of course, but it will take some time because I will have to correct your misconceptions and because it is non-trivial in the first place. I hope that the presented solution is acceptable to you, at least as a workaround for another two days, while I finish the work I am currently immersed in.

**Comment by joaotavora** [comment]
Alright, i understand why interned symbols are needed finally.  There are two keymaps with an indirection which is the symbol.  So, here's a better patch, not much more complicated:

```diff
diff --git a/lisp/transient.el b/lisp/transient.el
index f9060f5ba85..1ee19d9944f 100644
--- a/lisp/transient.el
+++ b/lisp/transient.el
@@ -1098,6 +1098,10 @@ transient--parse-group
                   (cl-mapcan (lambda (s) (transient--parse-child prefix s))
                              spec))))))

+(defvar transient-ginormous-interned-symbols nil)
+(defun transient--intern-maybe (name)
+  (if transient-ginormous-interned-symbols (intern name) (gensym "transient:")))
+
 (defun transient--parse-suffix (prefix spec)
   (let (level class args)
     (cl-symbol-macrolet
@@ -1127,7 +1131,7 @@ transient--parse-suffix
        ((and (commandp car)
              (not (stringp car)))
         (let ((cmd pop)
-              (sym (intern
+              (sym (transient--intern-maybe
                     (format "transient:%s:%s"
                             prefix
                             (let ((desc (plist-get args :description)))
@@ -1156,7 +1160,7 @@ transient--parse-suffix
              (when-let ((shortarg (transient--derive-shortarg arg)))
                (setq args (plist-put args :shortarg shortarg)))
              (setq args (plist-put args :argument arg))))
-          (setq sym (intern (format "transient:%s:%s" prefix arg)))
+          (setq sym (transient--intern-maybe (format "transient:%s:%s" prefix arg)))
           (setq args (plist-put
                       args :command
                       `(prog1 ',sym
```

Before the patch:

```
src/emacs -Q -f fido-mode -L ~/Source/Emacs/gptel -l gptel-transient -f gptel-menu
h
C-g
M-x vc SPC
```

Will bring up gibberish.

After the patch it won't.  You can customize `transient-ginormous-interned-symbols` to `t` to recover ginormous symbols.

> I think you misunderstand what my predicate does. 

No, I'm pretty sure I understand it fixes the problem.  For me (so thanks!).

But this isn't about me.  This is about preventing a library -- that isn't exactly an opt-out -- from polluting the obarray with commands that by definition don't need to be there except perhaps (but that depends on the implementation) for the time the transient menus are available.

in other words, this is a plain bug in my sight.  We mustn't tell users of the `flex` completion style to either customize a bug away or stay out of a given package.

Are you sure there's no way to do exactly what transient does today while keeping the API and avoiding this pollution.

> because it is non-trivial in the first place. I hope that the presented solution is acceptable to you, at least as a workaround for another two days, while I finish the work I am currently immersed in.

Sure, take your time, and don't hold back on the technical arguments.

**Comment by tarsius** [comment]
Unlike @psionic-k does at [68568](https://debbugs.gnu.org/cgi/bugreport.cgi?bug=68568), I agree with @joaotavora, that Transient should actively prevent the commands, that are only intended to be used from a transient prefix, from being presented as completion choices in `execute-extended-command`.

But I still think that the way to do that is by setting `read-extended-command-predicate`. I have further refined that in 522b625cf39cb1a9322719dcb039f25bc49ab35b. That completely removes the possibility that any other commands are accidentally affected by Transient's predicate. The commit message explains why I think that is the right way to do it, answering some of the above questions in the process.

**Comment by joaotavora** [comment]
@tarsius I didn't understand you meant to set this user-visible variable from transient.el's .  

I'm fairly sure there are other ways to go about this, but fair enough.  Anyway,  if you want to do this, shouldn't you be using `add-function` instead of `setq`?  Should be the go-to choice for this kinds of function-holding variables, and there's no if required.

```
(add-function :around read-extended-command-predicate #'funcall '((name . joaot/thingy)))   ;; dummy pass-through
```

**Comment by tarsius** [comment]
> shouldn't you be using `add-function` instead of `setq`?  Should be the go-to choice for this kinds of function-holding variables, and there's no if required.

I was vaguely aware of this being a thing, but have never done it before, and am unclear on when one should and should do that. Sounds like a good idea. (But it is more expensive.)

So I would do this, right?:

```elisp
(defun transient--command-completion-not-suffix-only-p (fn symbol buffer)
  "Return nil if SYMBOL is only useful as a transient suffix command.
Otherwise, if that is non-nil, call FN and return its value.
FN is the value of `read-extended-command-predicate'.
If that is nil, return t."
  (cond ((eq (get symbol 'completion-predicate) 'transient--suffix-only)
         nil)
        (fn (funcall fn symbol buffer))
        (t)))

(static-if (boundp 'read-extended-command-predicate)
    (add-function :around read-extended-command-predicate
                  #'transient--command-completion-not-suffix-only-p))
```

Hm, the Custom interface (and `describe-variable`) don't deal with this in a useful manner:

```
Hide Read Extended Command Predicate: Choice: Value Menu Other predicate function: 
#[128 "\304\300\301#\207"
  [transient--command-completion-not-suffix-only-p nil :around nil apply]
  5 advice]
    State : changed
```

I would have hoped that the user would be able to change the "core" value of the option, without discarding the advise. I.e., it probably would be better if `read-extended-command-predicate` were a hook.

**Comment by joaotavora** [comment]
> I would have hoped that the user would be able to change the "core" value of the option, without discarding the advise

I had exactly the same expectation. Maybe @monnier can shine some light here. Maybe it's not hard to fix...

> . I.e., it probably would be better if read-extended-command-predicate were a hook.

Yup. IIUC this is one of the things `add-function` is supposed to allow: make a mode-tweakable hook out of any old function-holding variable. But I might be mistaken

**Comment by monnier** [comment]
>> I would have hoped that the user would be able to change the "core" value
>> of the option, without discarding the advise
> I had exactly the same expectation. Maybe @monnier can shine some light
> here. Maybe it's not hard to fix...

The `C-h` output should show the `cl-print` pretty printed advice, which
is a bit better than the raw bytecode.

As for the user modifying this "core" value, while `nadvice.el` clearly
supports it for `advice-add` (via `defalias`), there isn't
a corresponding operation for `add-function`.
Instead, users would usually just *add* to the value with `:after-while`.

BTW, you can simplify your code to:

    (defun transient--command-completion-not-suffix-only-p (symbol buffer)
      (not (eq (get symbol 'completion-predicate) 'transient--suffix-only)))
    
    (static-if (boundp 'read-extended-command-predicate)
        (add-function :after-while read-extended-command-predicate
                      #'transient--command-completion-not-suffix-only-p))

BTW, I still don't see why those symbols need to be interned, so at the
very least, the code deserves a comment explaining why we have to do
that (and then have to do the `read-extended-command-predicate` dance)

[ Oh, BTW, the above code won't work because
  `read-extended-command-predicate` defaults to nil (i.e. its type is
  not "function" but (or function null)), which is an all too
  common mistake.  It should default to a function (e.g. `ignore`),
  otherwise `add-function` can't be used with it.  ]

**Comment by tarsius** [comment]
> BTW, I still don't see why those symbols need to be interned, so at the very least, the code deserves a comment explaining why we have to do that (and then have to do the `read-extended-command-predicate` dance)

I am glad you used the term "dance" here, because that's exactly how I feel about this request.

I have already said that it would be possible to avoid interning these symbols.  That is not actually true though -- but it would be possible to avoid making them fbound.

Every "suffix" consists of two parts: a command and an object. These two sides of the suffix have to somehow be linked to each other, and I have very early on decided that I would use a symbol to achieve that.

`(transient-suffix-object this-command)` and `(oref obj command)` are used to look up the other side of the suffix.  The object is stored in a property I made up for that purpose, `transient--suffix`.  For the command I chose to use the symbol's function slot.

However, because I wanted to avoid polluting `M-x`, I did initially use another slot for "anonymous" commands.  I was not happy with that from the start and only saw it as a temporary work-around -- a "dance" -- only to be used until a proper solution came along, which eventually did: `read-extended-command-predicate`.

No longer did Transient have to bend over backward to avoid polluting `M-x`.  As mentioned before, a suffix has to sides the command and the object, and they are linked via a symbol.  While the old M-x-avoidance hack was in place, a second dichotomy was necessary; the command existed in a "call-interactive-able" form and in a "symbol-for-identification" form. 

To get a symbol that can be used to locate the corresponding object, `this-command` had to be passed though this:

``` emacs-lisp
(defun transient--suffix-symbol (arg)
  "Return a symbol representing ARG.

ARG must be a command and/or a symbol.  If it is a symbol,
then just return it.  Otherwise return the symbol whose
`transient--infix-command' property's value is ARG."
  (or (cl-typep arg 'command)
      (cl-typep arg 'symbol)
      (signal 'wrong-type-argument `((command symbol) ,arg)))
  (if (symbolp arg)
      arg
    (let* ((obj (transient-suffix-object))
           (sym (oref obj command)))
      (if (eq (get sym 'transient--infix-command) arg)
          sym
        (catch 'found
          (mapatoms (lambda (sym)
                      (when (eq (get sym 'transient--infix-command) arg)
                        (throw 'found sym)))))))))
```

Nearly all commands that change the value of some argument can be implemented exactly the same:

``` emacs-lisp
(lambda ()
  (interactive)
   (let ((obj (transient-suffix-object)))
     (transient-infix-set obj (transient-infix-read obj)))
   (transient--show))
```

But for that to work `transient-suffix-object` has to be able to determine the corresponding object.  With the current implementation that is done with `(get this-command 'transient--object)`.  Previously it had to pass `this-command` through the `transient--suffix-symbol` function shown above, to map through all symbols to find the one whose `transient--infix-command` property is `eq` to `this-command`.  This also meant that even-though all those commands were `equal`, they could not be `eq`, leading to a lot of wasted memory.

I think adding a little function to `read-extended-command-predicate` is a better approach.  It's easier to reason about, doesn't have to loop over all symbols to find the correct one, and avoids wasting memory.

But lets assume for a moment that we go back to the old kludge anyway.  Some infix argument commands are not defined anonymously within the definition of a prefix command, because they are shared between multiple prefixes.  For example:

``` emacs-lisp
(transient-define-argument magit:--gpg-sign ()
  :description "Sign using gpg"
  :class 'transient-option
  :shortarg "-S"
  :argument "--gpg-sign="
  :allow-empty t
  :reader #'magit-read-gpg-signing-key)
```

This is used by `magit-commit`, `magit-merge`, `magit-rebase` and a few other prefixes, but when called outside the context of a prefix, it is useless, just like the infixes defined anonymously for a single prefix.  So this too should be excluded from `M-x` -- and how do we do that? We use an appropriate `read-extended-command-predicate`.  And if we have to do that anyway to hide a bunch of named commands, can't we take advantage of that to hide the anonymously defined commands as well?

Finally, I don't think that Transient defining commands that are only useful in the context of a *certain transient prefix command* (and nevertheless putting these commands in the global namespace), is all that different, from a package defining commands that are only useful in the context a *certain major-mode* (and nevertheless putting these commands in the global namespace).  If it is acceptable for these major-mode specific commands to "pollute M-x", then shouldn't it be acceptable for the transient-prefix specific commands to do the same?  The pollution is a problem -- in both cases -- and `read-extended-command-predicate` helps addressing it -- again in both cases.

**Comment by monnier** [comment]
> I am glad you used the term "dance" here, because that's exactly how
> I feel about this request.

OK, reading&thinking some more, we shouldn't need to fiddle with the
global `read-extended-command-predicate`.  Instead you should be able to
set your symbols' `completion-predicate` properties to something like `ignore`.

> Every "suffix" consists of two parts: a command and an object. These
> two sides of the suffix have to somehow be linked to each other, and
> I have very early on decided that I would use a symbol to
> achieve that.

[ Maybe I'm suffering from the "man with a hammer" syndrome, but the above
  words make me think that maybe what you want is an OClosure,
  i.e. something that's both a command and an object at the same time.  ]

> `(transient-suffix-object this-command)` and `(oref obj command)` are
> used to look up the other side of the suffix.  The object is stored in
> a property I made up for that purpose, `transient--suffix`.  For the
> command I chose to use the symbol's function slot.
>
> However, because I wanted to avoid polluting `M-x`, I did initially
> use another slot for "anonymous" commands.  I was not happy with that
> from the start and only saw it as a temporary work-around --
> a "dance" -- only to be used until a proper solution came along, which
> eventually did: `read-extended-command-predicate`.

That still doesn't explain why you need them to be interned.

> No longer did Transient have to bend over backward to avoid polluting
> `M-x`.  As mentioned before, a suffix has to sides the command and the
> object, and they are linked via a symbol.  While the old M-x-avoidance
> hack was in place, a second dichotomy was necessary; the command
> existed in a "call-interactive-able" form and in
> a "symbol-for-identification" form. 
>
> To get a symbol that can be used to locate the corresponding object,
> `this-command` had to be passed though this:
>
> ``` emacs-lisp
> (defun transient--suffix-symbol (arg)
>   "Return a symbol representing ARG.
>
> ARG must be a command and/or a symbol.  If it is a symbol,
> then just return it.  Otherwise return the symbol whose
> `transient--infix-command' property's value is ARG."
>   (or (cl-typep arg 'command)
>       (cl-typep arg 'symbol)
>       (signal 'wrong-type-argument `((command symbol) ,arg)))
>   (if (symbolp arg)
>       arg
>     (let* ((obj (transient-suffix-object))
>            (sym (oref obj command)))
>       (if (eq (get sym 'transient--infix-command) arg)
>           sym
>         (catch 'found
>           (mapatoms (lambda (sym)
>                       (when (eq (get sym 'transient--infix-command) arg)
>                         (throw 'found sym)))))))))
> ```

OK, this would indeed require the symbol to be interned for `mapatoms`
to find it, but AFAICT you don't use that code any more.

> But for that to work `transient-suffix-object` has to be able to
> determine the corresponding object.  With the current implementation
> that is done with `(get this-command 'transient--object)`.

But (get this-command 'transient--object) should work as well whether
`this-command` holds an interned symbol or non-interned one.

> Finally, I don't think that Transient defining commands that are only
> useful in the context of a *certain transient prefix command* (and
> nevertheless putting these commands in the global namespace), is all
> that different, from a package defining commands that are only useful
> in the context a *certain major-mode* (and nevertheless putting these
> commands in the global namespace).

Indeed.  There's nothing fundamentally wrong with it.

It's just more often inconvenient for some end-users, maybe because of
the number of those commands, or their length (which makes it more
likely that they'll match some search pattern).

**Comment by tarsius** [comment]
Okay. I'll experiment.
